Content.Shared/Inventory/InventorySystem.Equip.cs

```diff
@@ -0,0 +1,244 @@
+    }
+
+    private void OnEntRemoved(EntityUid uid, InventoryComponent component, EntRemovedFromContainerMessage args)
+    {
+        if(!TryGetSlot(uid, args.Container.ID, out var slotDef, inventory: component))
+            return;
+
+        var unequippedEvent = new DidUnequipEvent(uid, args.Entity, slotDef);
+        RaiseLocalEvent(uid, unequippedEvent);
+
+        var gotUnequippedEvent = new GotUnequippedEvent(uid, args.Entity, slotDef);
+        RaiseLocalEvent(args.Entity, gotUnequippedEvent);
+    }
+
+    private void OnEntInserted(EntityUid uid, InventoryComponent component, EntInsertedIntoContainerMessage args)
+    {
+        if(!TryGetSlot(uid, args.Container.ID, out var slotDef, inventory: component))
+           return;
+
+        var equippedEvent = new DidEquipEvent(uid, args.Entity, slotDef);
+        RaiseLocalEvent(uid, equippedEvent);
+
+        var gotEquippedEvent = new GotEquippedEvent(uid, args.Entity, slotDef);
+        RaiseLocalEvent(args.Entity, gotEquippedEvent);
+    }
+
+    public bool TryEquipActiveHandTo(EntityUid uid, string slot, bool silent = false, bool force = false,
+        InventoryComponent? component = null, SharedHandsComponent? hands = null)
+    {
+        if (!Resolve(uid, ref component, false) || !Resolve(uid, ref hands, false))
+            return false;
+
+        if (!hands.TryGetActiveHeldEntity(out var heldEntity))
+            return false;
+
+        return TryEquip(uid, heldEntity, slot, silent, force, component);
+    }
+
+    public virtual bool TryEquip(EntityUid uid, EntityUid itemUid, string slot, bool silent = false, bool force = false, InventoryComponent? inventory = null, SharedItemComponent? item = null)
+    {
+        if (!Resolve(uid, ref inventory, false) || !Resolve(itemUid, ref item, false))
+        {
+            if(!silent) _popup.PopupCursor(Loc.GetString("inventory-component-can-equip-cannot"), Filter.Local());
+            return false;
+        }
+
+        if (!TryGetSlotContainer(uid, slot, out var slotContainer, out var slotDefinition, inventory))
+        {
+            if(!silent) _popup.PopupCursor(Loc.GetString("inventory-component-can-equip-cannot"), Filter.Local());
+            return false;
+        }
+
+        if (!force && !CanEquip(uid, itemUid, slot, out var reason, slotDefinition, inventory, item))
+        {
+            if(!silent) _popup.PopupCursor(Loc.GetString(reason), Filter.Local());
+            return false;
+        }
+
+        if (!slotContainer.Insert(itemUid))
+        {
+            if(!silent)  _popup.PopupCursor(Loc.GetString("inventory-component-can-unequip-cannot"), Filter.Local());
+            return false;
+        }
+
+        if(item.EquipSound != null)
+            SoundSystem.Play(Filter.Pvs(uid), item.EquipSound.GetSound(), uid, AudioParams.Default.WithVolume(-2f));
+
+        inventory.Dirty();
+
+        _movementSpeed.RefreshMovementSpeedModifiers(uid);
+
```

Comment:
shouldn't the movement speed system just subscribe to did (un)equip events?

Reply:
In an ideal world. Pretty sure it was coded before the eventbus was even a thing.

Reply:
Yeah previously it probably made sense. But we are steadily moving towards an ideal world so might as well change this now.